akka-stream-testkit/src/main/scala/akka/stream/testkit/StreamTestKit.scala

```diff
@@ -0,0 +1,443 @@
+   */
+  class Probe[T] private[TestPublisher] (initialPendingRequests: Long)(implicit system: ActorSystem) extends ManualProbe[T] {
+
+    type Self = Probe[T]
+
+    private var pendingRequests = initialPendingRequests
+    private lazy val subscription = expectSubscription()
+
+    /**
+     * Current pending requests.
+     */
+    def pending: Long = pendingRequests
+
+    def sendNext(elem: T): Self = {
+      if (pendingRequests == 0) pendingRequests = subscription.expectRequest()
+      pendingRequests -= 1
+      subscription.sendNext(elem)
+      this
+    }
+
+    def unsafeSendNext(elem: T): Self = {
+      subscription.sendNext(elem)
+      this
+    }
+
+    def sendComplete(): Self = {
+      subscription.sendComplete()
+      this
+    }
+
+    def sendError(cause: Exception): Self = {
+      subscription.sendError(cause)
+      this
+    }
+
+    def expectRequest(): Long = subscription.expectRequest()
+
+    def expectCancellation(): Self = {
+      subscription.expectCancellation()
+      this
+    }
+  }
+
+}
+
+object TestSubscriber {
+
+  import StreamTestKit._
+
+  /**
+   * Probe that implements [[org.reactivestreams.Subscriber]] interface.
+   */
+  def manualProbe[T]()(implicit system: ActorSystem): ManualProbe[T] = new ManualProbe()
+
+  def probe[T]()(implicit system: ActorSystem): Probe[T] = new Probe()
+
+  /**
+   * Implementation of [[org.reactivestreams.Subscriber]] that allows various assertions.
+   */
+  class ManualProbe[I] private[TestSubscriber] ()(implicit system: ActorSystem) extends Subscriber[I] {
+
+    type Self <: ManualProbe[I]
+
+    private val probe = TestProbe()
+
+    private val self = this.asInstanceOf[Self]
+
+    /**
+     * Expect and return a Subscription.
+     */
+    def expectSubscription(): Subscription = probe.expectMsgType[OnSubscribe].subscription
+
+    /**
+     * Expect [[SubscriberEvent]].
+     */
+    def expectEvent(event: SubscriberEvent): Self = {
+      probe.expectMsg(event)
+      self
+    }
+
+    /**
+     * Expect a data element.
+     */
+    def expectNext(element: I): Self = {
+      probe.expectMsg(OnNext(element))
+      self
+    }
+
+    /**
+     * Expect multiple data elements.
+     */
+    def expectNext(e1: I, e2: I, es: I*): Self = expectNextN(e1 +: e2 +: es)
+
+    def expectNextUnordered(e1: I, e2: I, es: I*): Self = expectNextUnorderedN(e1 +: e2 +: es)
+
+    /**
+     * Expect and return a data element.
+     */
+    def expectNext(): I = probe.expectMsgType[OnNext[I]].element
+
+    def expectNextN(all: Seq[I]): Self = {
+      all.foreach(e ⇒ probe.expectMsg(OnNext(e)))
+      self
+    }
+
+    def expectNextUnorderedN(all: Seq[I]): Self = {
+      @annotation.tailrec def expectOneOf(all: Seq[I]): Unit = all match {
+        case Nil ⇒
+        case list ⇒
+          val next = expectNext()
+          assert(all.exists(_ == next), s"expected one of $all, but received $next")
+          expectOneOf(all.filterNot(_ == next))
```

Comment:
Yes, I'd vote for adjusting impl to work like you suggest (didn't review in full before just noticed your comment above)